lib/connectors/memory.js

```diff
@@ -352,29 +352,61 @@ Memory.prototype.fromDb = function(model, data) {
   if (!data) return null;
   data = deserialize(data);
   var props = this._models[model].properties;
+
   for (var key in data) {
     var val = data[key];
-    if (val === undefined || val === null) {
-      continue;
-    }
+
     if (props[key]) {
-      switch (props[key].type.name) {
-        case 'Date':
-          val = new Date(val.toString().replace(/GMT.*$/, 'GMT'));
-          break;
-        case 'Boolean':
-          val = Boolean(val);
-          break;
-        case 'Number':
-          val = Number(val);
-          break;
-      }
+      val = this._castPropertyValue(key, val, props);
     }
+
     data[key] = val;
   }
+
   return data;
 };
 
+Memory.prototype._castPropertyValue = function(prop, val, props) {
+  if (val === undefined || val === null) {
+    return val;
+  }
+
+  if (Array.isArray(val)) {
+    var self = this;
+    return val.map(function(val) {
+      return self._castPropertyValue(prop, val, props);
+    });
+  }
+
+  var isArray = Array.isArray(props[prop].type);
+  var name = isArray ? props[prop].type[0].name : props[prop].type.name;
+
+  switch (name) {
+    case 'Date':
+      val = new Date(val.toString().replace(/GMT.*$/, 'GMT'));
+      break;
+    case 'Boolean':
+      val = Boolean(val);
+      break;
+    case 'Number':
+      val = Number(val);
+      break;
+    case 'ModelConstructor':
+      for (var subProp in val) {
+        val[subProp] = isArray ?
+          this._castPropertyValue(
+            subProp,
+            val[subProp],
+            props[prop].type[0].definition.properties
```

Comment:
The logic of choosing between `props[prop].type` vs. `props[prop].type[0]` is duplicated. Please extract the item type only once.

Something like this looks best to me:

```js
var isArray = ...
if (isArray) {
  propType = propType[0];
}

var name = propType.name;
switch (...) {
  // ...
  case 'ModelConstructor':
    for (var subProp in val) {
      this._castPropertyValue(subProp, val[subProp], propType.definition.properties);
    }
}
```